### ISSUE TITLE
Content off screen and keyboard opens when navigating to default rooms 

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#8292]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create workspace 
2. Navigate to default room 

## Expected Result:
Navigating to a default room shows the chat content within the page

## Actual Result:
Navigating to a default room shows the chat content half off screen and the keyboard opens
Issue is repro when no chats in the room 

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Mobile Web - Safari

**Version Number:** 1.1.44-3
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
![IMG_0158D0729524-1](https://user-images.githubusercontent.com/43995119/159851361-d8a6e039-9187-4960-8e66-71acc1e1adcb.jpeg)

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @neil-marcellini 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1647980292349759

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Beamanator** [comment]
Looks like a good issue to be solved externally 👍

**Comment by adelekennedy** [comment]
[Internal](https://www.upwork.com/ab/applicants/1507512096161005568/job-details)
[External](https://www.upwork.com/jobs/~01a87101625eccc331)

**Comment by adelekennedy** [comment]
And it's time to double again

**Comment by adelekennedy** [comment]
It's time to double again!

**Comment by anboia** [comment]
## Proposal

Remove the autoFocus, when the room has just been created (no action history)

## Overview

What is causing this issue, is when you open a room, and there is no action on that report, except the "CREATED" `actionName`, this enables autoFocus

https://github.com/Expensify/App/blob/e7565c1aae9ee4a8f366932d63483619d742e970/src/pages/home/report/ReportActionCompose.js#L507

as you can see in this line, the autoFocus is being tiggered by `_.size(this.props.reportActions) === 1` condition.

## Solution

Removes the `_.size(this.props.reportActions) === 1` condition, this will open the report screen without autofocusing the text input, and the issue will be gone.

This will change the behavior of having the autofocus when you open a room without any previous action. but it will keep consistency with other rooms that have already an action history (list of messages).

@tgolen

**Comment by tgolen** [comment]
OK, I think we want to still have the auto-focus functionality. That was purposefully added. I think the solution for this is to fix why the content is halfway off the screen.

**Comment by rushatgabhane** [comment]
@adelekennedy not moving forward with @anboia's propsoal because https://github.com/Expensify/App/issues/8292#issuecomment-1108866564

**Comment by rushatgabhane** [comment]
@Expensify/contributor-plus feel free to pick this issue up, I'm unassigning because I'll be OOO.
No untriaged proposals yet. Thanks!

**Comment by parasharrajat** [comment]
https://jsbin.com/kefisiceme/edit?html,css,js

**Comment by michaelhaxhiu** [comment]
note: I am re-assigning @parasharrajat because he's already started to assess proposals.

**Comment by parasharrajat** [comment]
@anboia any thoughts on https://github.com/Expensify/App/issues/8292#issuecomment-1110242829?

**Comment by anboia** [comment]
### Proposal

Removing a call for `setNativeProps` function on `InvertedFlatList`, fixes the issue

https://github.com/Expensify/App/blob/11a708c5614574e24f714de30f59683205e8d530/src/components/InvertedFlatList/index.js#L44-L48

`ReportActionsList` is the only component that uses `InvertedFlatList`. I have tested it on all web platforms and worked great.

### Context

According to necolas/react-native-web#1935 `setNativeProps` should be avoided. It overwrites the previous style, since the style is being updated from the animation, and at the same time, the keyboard is showing up (focus effect) the `setNativeProps` is not being effective. 

@parasharrajat

**Comment by parasharrajat** [comment]
Did you get a chance to go through CONTRIBUTING.md? I am sure it was mentioned on the Upwork Job. Please close that PR and do not create any until your proposal is accepted.

**Comment by anboia** [comment]
> Did you get a chance to go through CONTRIBUTING.md? I am sure it was mentioned on the Upwork Job. Please close that PR and do not create any until your proposal is accepted.

Ok, sorry about that

**Comment by anboia** [comment]
Maybe it will be the best approach, manually setting the the focus with a timeout?... I just noticed the twinkled, very sutil but exist. And doesn't seams to be an option to apply the style in another way because it seams that the style is already being handled by the react-native-web (on vitualizedList) when inverted={true}

**Comment by parasharrajat** [comment]
Ok, let's us know how would you do that.

**Comment by adelekennedy** [comment]
Waiting on updated proposal

**Comment by adelekennedy** [comment]
still waiting @anboia any movement here?

**Comment by parasharrajat** [comment]
I am open to proposals.

**Comment by anboia** [comment]
### Proposal

instead of using `autoFocus`, manually focus using the `focus(true)` function. Passing 'true' will delay the focus by 100ms, this function is also used to focus, after closing the emoji modal. 

<img width="947" alt="image" src="https://user-images.githubusercontent.com/4477665/168635678-cc51b0f7-ab02-4996-b5e1-43738dd12ca8.png">

there is only one downside here, calling `focus()` does not shows the keyboard on the iOS web. but this seems to be a different issue, that might need to be handled or not. the comments explain that is using a setTimeout as a workaround, but that workaround does not seams to be working.

https://github.com/Expensify/App/blob/5c5f79b86f716c8187fd2ad9c0dd0af743e68bd1/src/pages/home/report/ReportActionCompose.js#L314-L322

I think we can go this way, since the half screen will be fixed (by applying https://github.com/Expensify/App/issues/8292#issuecomment-1120272037), and it will get the focus. @parasharrajat

**Comment by anboia** [comment]
The problem with `autoFocus` is also that, it does not wait for the screen navigation ends. Once an element gets focused, on ios, the elements get visible, skipping the animation end, as @parasharrajat mentioned here https://github.com/Expensify/App/issues/8292#issuecomment-1110242829

With manual focus, we can trigger the focus only after animation (`InteractionManager.runAfterInteractions`)

**Comment by rushatgabhane** [comment]
@anboia @parasharrajat just dropping a note for you to test this case too https://expensify.slack.com/archives/C01GTK53T8Q/p1652813754277019

**Comment by adelekennedy** [comment]
not overdue - still dialing in proposals

**Comment by adelekennedy** [comment]
doubling time!

**Comment by anboia** [comment]
### Proposal
## iOS constraints:
* The only way to bring up the keyboard when using manual `.focus()`, is by calling `.focus()` inside a click handler triggered by user interaction. (can't find documentation for that, but this is mentioned in most related issues)
* It does not work if the focus is called async
* It is possible to switch focus to your target input (async way) when some other input is already focused

## Solution
We can create a fake input element, and call `.focus()` on that input, before calling `.focus()` on the actual `Composer` text input async. Have tested it, and it works, without flickering

### Solution details:
* Add a call to `ReportActionComposeFocusManager.focus()` inside the `onLinkClick` function:
https://github.com/Expensify/App/blob/8e3c67ba25200f5558ce6be6fd3aa211c43d60b6/src/pages/home/sidebar/SidebarScreen.js#L111

* create a fake input element, to syncrhonusly call `fakeInput.focus()` on that element. (https://stackoverflow.com/a/55652503)
```javascript
const fakeInput = document.createElement('input');
fakeInput.setAttribute('type', 'text');
fakeInput.style.position = 'absolute';
fakeInput.style.opacity = 0;
fakeInput.style.height = 0;
fakeInput.style.fontSize = '16px'; // disable auto zoom
document.body.prepend(fakeInput);
```
* add a call to `fakeInput.focus()` before calling `textInput.focus()` on the `Composer` text input asyncrounsly.


https://user-images.githubusercontent.com/4477665/170223258-6f24c636-568e-4613-b5b2-301b21cc7172.mp4


@parasharrajat

**Comment by anboia** [comment]
here are the changes compare: https://github.com/Expensify/App/compare/main...anboia:fix/half-inverted-list?expand=1

**Comment by parasharrajat** [comment]
Hmm, we are familiar with this hack and don't want to proceed with this hack unless necessary. 

I will discuss the previous proposal internally to see if others agree to do that.

**Comment by parasharrajat** [comment]
Hmm... No new proposals. I will open up the discussion on Monday and let's see what others agree to.

**Comment by adelekennedy** [comment]
Not overdue - still discussing next steps

**Comment by parasharrajat** [comment]
Asked about it here https://expensify.slack.com/archives/C02NK2DQWUX/p1654017861456809

**Comment by anboia** [comment]
> Asked about it here https://expensify.slack.com/archives/C02NK2DQWUX/p1654017861456809

Hey @parasharrajat , The link doesn't open for me. Any update on this? Thanks

**Comment by parasharrajat** [comment]
Sorry, no updates yet.

**Comment by mollfpr** [comment]
https://github.com/react-navigation/react-navigation/issues/10014
https://github.com/react-navigation/react-navigation/issues/9875

Might have the same problem with those. By the way, this issue also happens on my iOS Chrome with the staging site.

I have a minimum workaround for this. I notice that on the iOS app, there's a loading indicator showing up before showing the `ReportActionsView`. Maybe we could show the loader too before showing the `ReportActionCompose`?

https://github.com/Expensify/App/blob/fd29904ef9857e6c5b858da04d52c8dec0d5c369/src/pages/home/ReportScreen.js#L142
```diff
-  this.loadingTimerId = setTimeout(() => this.setState({isLoading: false}), 0);
+ this.loadingTimerId = setTimeout(() => this.setState({isLoading: false}), _.size(this.props.reportActions) === 1 ? 500 : 0);
```
Like in here we could add more time when trying to open a new report chat. We could add just a `100` timeout but the cursor is not aligned properly when the input is loaded.

https://github.com/Expensify/App/blob/fd29904ef9857e6c5b858da04d52c8dec0d5c369/src/pages/home/ReportScreen.js#L192
```diff
-  {(isArchivedRoom || this.props.session.shouldShowComposeInput) && (
+  {!this.shouldShowLoader() && (isArchivedRoom || this.props.session.shouldShowComposeInput) && (
```
And in here we wait till the loader is done before showing the composer.

### With 500ms

https://user-images.githubusercontent.com/25520267/171921097-55b3d03c-30af-4884-9001-67ea209e515e.mp4

### With 100ms

https://user-images.githubusercontent.com/25520267/171921288-2799b115-8612-4fb2-b329-2f5af1d00b1c.mp4

**Comment by mdneyazahmad** [comment]
## Proposal

As the root cause is already understood, as soon as the component is mounted, it tries to focus and it causes the issue. I am proposing to emit an event `drawerClosed` in `@react-nativation/drawer`. We then focus when the drawer is closed.

https://github.com/Expensify/App/blob/fd29904ef9857e6c5b858da04d52c8dec0d5c369/src/pages/home/ReportScreen.js#L203-L208

Pass `navigation` to `ReportActionCompose`

```diff
 <ReportActionCompose 
     onSubmit={this.onSubmitComment} 
     reportID={reportID} 
     reportActions={this.props.reportActions} 
     report={this.props.report} 
+   navigation={this.props.navigation}
 /> 
```

Add event listener (`drawerClosed`) and focus when called and unsubscribe. Because we want to focus on it only one time.

```javascript
        // file:  src/pages/home/report/ReportActionCompose.js

        const unsubscribe = this.props.navigation.addListener('drawerClosed', () => {
            if (this.shouldFocusInputOnScreenFocus || _.size(this.props.reportActions) === 1) {
                this.textInput.focus();
            }

            unsubscribe();
        });
```

```diff
                                <Composer
-                                   autoFocus={this.shouldFocusInputOnScreenFocus || _.size(this.props.reportActions) === 1}
```

### `@react-nativation/drawer`

It does not have `drawerClosed` event. so we need to add this feature. Here's the code (for testing) that adds this feature. (May not be perfect, I will invest more time if the proposal sounds good.).

There are two implementation of Drawer component legacy and modern. During my testing it seems to use legacy.

```diff
  // file: App/node_modules/@react-navigation/drawer/lib/module/views/legacy/Drawer.js

 _defineProperty(this, "transitionTo", isOpen => {

       ...

      stopClock(this.clock), call([this.isOpen], _ref => {
        let [value] = _ref;
        const open = Boolean(value);
        this.handleEndInteraction();

+        this.props.navigation.emit({
+          type: 'drawerClosed',
+        });

        if (open !== this.props.open) {
```

```diff
    // file: App/node_modules/@react-navigation/drawer/lib/module/views/DrawerView.js

  return /*#__PURE__*/React.createElement(DrawerStatusContext.Provider, {
    value: drawerStatus
  }, /*#__PURE__*/React.createElement(Drawer, {

    ...

    renderDrawerContent: renderDrawerContent,
    renderSceneContent: renderSceneContent,
    dimensions: dimensions,
+  navigation: navigation,

```

**Comment by parasharrajat** [comment]
@mdneyazahmad I got your idea. Although you didn't explain it your proposal that why it works, I think it can a solution to the problem. But I doubt that your approach will be ideal. Did you check any old issues on React-navigation about requests for such (drawerClosed) events? Let's see them. 

I guess author does not consider this event necessary, that's why he didn't implemented it. Did you explore existing API which can be used for `drawerClosed` events? 

> There are two implementation of Drawer component legacy and modern. During my testing it seems to use legacy.

If you are changing the API of a lib then it must be implemented in all variants i.e. Legacy and modern both. 

You code snippets use build lib files, instead of this try to show changes in Source files(also for future issues, for any project). 

Could you please show an video for the effect of your changes? It would be better as code changes are third party.

**Comment by mollfpr** [comment]
### Proposal

I think this issue is related to CSS animation on the browser. I've seen some issue on StackOverflow the animation change behavior when focusing the input at the same time.
https://stackoverflow.com/questions/54075497/example-of-input-with-focus-breaking-slide-down-animation-of-form
https://stackoverflow.com/questions/24742930/using-css-animations-on-element-containing-a-focused-input-box

My solution is change the `drawerType` from `slide` to `front` for mWeb.

https://user-images.githubusercontent.com/25520267/173235503-568d749b-4598-49cd-ac84-3f34c61585fd.mov

**Comment by mdneyazahmad** [comment]
>  I got your idea. Although you didn't explain it in your proposal that why it works,

If an animation is on going and we try to focus on an element inside. It does not have consistent behavior. So, my proposal was to delay the focus until the animation completes.

> I guess author does not consider this event necessary, that's why he didn't implemented it. Did you explore existing API which can be used for drawerClosed events?

I have found an issue with similar feature request https://github.com/react-navigation/react-navigation/issues/10034 and the author has commented to use another method https://github.com/react-navigation/react-navigation/issues/10034#issuecomment-940864339. Following the docs I was able to fire a function when drawer is closed(rn-tester) but this does not work on our app may be because of custom actions we are resetting the state.

```javascript
import { getDrawerStatusFromState } from '@react-navigation/drawer';
import { useEffect } from 'react';

export default function AttachDrawerEvents({ navigation }) {
  useEffect(() => {
    return navigation.addListener('state', () => {
      const drawerState = getDrawerStatusFromState(navigation.getState());

      alert(drawerState);
    });
  }, [navigation]);

  return null;
}

```

https://user-images.githubusercontent.com/77761491/173336323-578c5f73-341c-46fa-b310-773d98a635e2.mp4

> If you are changing the API of a lib then it must be implemented in all variants i.e. Legacy and modern both.

Yes, It should implement all variants. As this was a proposal therefore I shared the idea and the final PR(if) will implement both.

> You code snippets use build lib files, instead of this try to show changes in Source files(also for future issues, for any project).

As this was proposal therefore I posted this (better to add the source too) so that you can quickly test it. As the source code should be compiled and linked to app may take some time.

Here is the source changes https://github.com/react-navigation/react-navigation/compare/main...mdneyazahmad:feat/drawer-events. But this has some issues like `drawerClosed` is fired twice and the legacy implementation here does not work need to fix that too. This code changes is slightly different than the original changes I proposed (for legacy).

> Could you please show an video for the effect of your changes? It would be better as code changes are third party.


https://user-images.githubusercontent.com/77761491/173336340-f5b8a41f-6e39-4c8c-9589-c9ecadae3c10.mp4

cc: @parasharrajat

**Comment by parasharrajat** [comment]
looking into the updates..

**Comment by parasharrajat** [comment]
@mdneyazahmad Good lead. As said in the comment, drawerClose event was removed from the react-navigation or it never existed. But there are other ways to determine the drawerStatus and ways to hook that into react ecosystem.  `state` event should be fired whenever the state changes and I see no reason why it is not firing. It may be possible that your subscription is recreated on each render.  

So I would suggest looking into the lib docs and finding the ways to listen to drawer close. I feel like adding `drawerClose` event is unnecessary and it will never be accepted into `react-navigation` API.

**Comment by aneequeahmad** [comment]
### PROPOSAL

_**PROBLEM**_

After debugging i found that the input field is `auto-focused` before the component is mounted(Drawer state is changed and other stuff is done). And that's why there is this `content-off screen` issue.

_**SOLUTION**_

As we only `autoFocus` for `touch-screen` devices. A new state variable `shouldAutoFocus` (boolean) is used and it is set to `true` in `componentDidMount` when the following condition is true. The piece of code is around `line# 154` in file`ReportActionCompose.js`.

```
         // Set autofocus on touch devices when input is mounted to fix the content off from screen
         if(this.shouldFocusInputOnScreenFocus || _.size(this.props.reportActions) === 1) {
            this.setState({shouldAutoFocus: true})
        }
```
and using this state variable in `<Composer>` as

```
          <Composer
              autoFocus={this.state.shouldAutoFocus}
              multiline
              ref={this.setTextInputRef}
              textAlignVertical="top"                 
  ```

**Comment by parasharrajat** [comment]
I do not understand your proposal @aneequeahmad. 

> As we only autoFocus for touch-screen devices.

How do you know this?

> After debugging i found that the input field is auto-focused before the component is mounted

Which component? How can a input be autofocused if it is not mounted?

I don't think autoFocus has any effect of state changes. It is used at `componentDidMount` only. Changing its value value `componentDidMount` should not trigger autoFocus. Am I missing something?

**Comment by aneequeahmad** [comment]
Also this issue isn't only `mWeb - Safari` specific. It can also be reproduced on `mWeb - Chrome`.

**Comment by aneequeahmad** [comment]
@parasharrajat you also see the following comments in `componentDidUpdate`. Just for more clarity. The main solution is to focus the input when all the stuff is done like `animation, screen is focused etc`

```
        // We want to focus or refocus the input when a modal has been closed and the underlying screen is focused.
        // We avoid doing this on native platforms since the software keyboard popping
        // open creates a jarring and broken UX.
        if (this.shouldFocusInputOnScreenFocus && this.props.isFocused
            && prevProps.modal.isVisible && !this.props.modal.isVisible) {
            this.focus();
        }
```

**Comment by mdneyazahmad** [comment]
@parasharrajat I also think the same that those events will not be accepted. But we have a way to know when the drawer is closed but that does not seem to work on our app. I tested it on the example app (react-navigation) it works there.

A state change event is fired always but it does not fire when we open a report. If we open the same report twice then it fires.


https://user-images.githubusercontent.com/77761491/173584705-7cbeadee-f9c2-4c5f-b2ef-aaee110910ea.mp4

I am looking into it, I will post as soon as I find the reason.

**Comment by aneequeahmad** [comment]
@parasharrajat anything still wrong in my proposal ? Doesn't i identified the root cause ?

**Comment by mdneyazahmad** [comment]
@aneequeahmad When an animation is ongoing and you try to focus an element inside. Its' behavior is not consistent, it breaks the animation. This is the root cause and we are trying to find a solution such that we can focus at the exact time when the animation is complete and not looking for a workaround with `setTimeout`. I do not think that your proposal https://github.com/Expensify/App/issues/8292#issuecomment-1154685173 will work as said @parasharrajat is doing a test to confirm.

**Comment by mdneyazahmad** [comment]
@parasharrajat I have found the issue. The component is unmounted every time the report is changed so all subscriptions are also canceled.

There is some inconsistency between legacy and modern implementation of drawer. As said our app uses legacy implementation.

Modern Implementation:
- As soon as the open drawer action is dispatched, Its drawer state is immediately set to open and the animation starts.
- And when the close drawer action is dispatched, an animation is triggered and when animation is complete it's drawer status is set to close.

Legacy Implementation:
- It always set drawer status to the state and then animation starts.

#### Why it does not work in our app:
When we navigate to a different report, it' s drawer state is immediately set to closed and now the animation starts (legacy implementation) and the state change subscription is added after the state is already state to closed, thus it does not invoke the subscription. This is not the case with modern implementation because there the animation starts first and when it completes then it fires the event and during this time state changed subscription is already ready to listen.

#### Solution:
Fix the inconstancy between modern and legacy implementation. Set the status to close after the animation complete (as in modern). This will be a upstream fix along with adding subscription to listen for this event and focus on our app.

**Comment by aneequeahmad** [comment]
@mdneyazahmad You're right about the root cause this is what i have said [here](https://github.com/Expensify/App/issues/8292#:~:text=drawer%20is%20closed%20but%20there%20is%20animation%20running%20that%20is%20not%20completed%20yet%20it%20needs%20some%20delay%20before%20focus%20is%20set). 

I haven't proposed that `setTimeout` workaround should be used. I said `setTimeout` is already used in the code as a `workaround` [here](https://github.com/Expensify/App/issues/8292#:~:text=//%20SetTimeout%20is%20used%20as%20a%20workaround).

When the animation completes the component is mounted and that is the point where we should focus the input. Here is the video that shows my proposal works.

https://user-images.githubusercontent.com/36359331/173694989-4ecd5759-3495-455d-84c7-3ee6cbce4b58.mp4

cc: @parasharrajat

**Comment by aneequeahmad** [comment]
@parasharrajat, I don't understand why we need state of drawer(closed) to focus the input. Isn't `componentDidMount` more reliable to use to allow the input to have focus ? 

Also, we can use `ref` to focus the input if we don't want to use `state variable` to set focus. We can simply use this piece of code to set focus. 

```
 // Set autofocus on touch devices when input is mounted to fix the content off from screen
         if(this.shouldFocusInputOnScreenFocus || _.size(this.props.reportActions) === 1) {
            this.textInput.current && this.textInput.current.focus();
        }
```

**Comment by parasharrajat** [comment]
> I don't understand why we need state of drawer(closed) to focus the input.

I am following the proposal and helping with queries. 

> Also, we can use ref to focus the input if we don't want to use the state variable to set focus. We can simply use this piece of code to set focus.

I will test your proposal as well. Give me some time to process it. Thanks for your patience.

**Comment by parasharrajat** [comment]
Need to review one proposal above. I will look into that tomorrow morning CST.

**Comment by aneequeahmad** [comment]
@parasharrajat any update on my proposal ? Have you fixed this issue ?

**Comment by parasharrajat** [comment]
No, I haven't fixed this. I will review your proposal as well.

**Comment by aneequeahmad** [comment]
Oops sorry my bad. I wanted to say have you tested that my proposal fixes the issue. Thankyou for your comment 🙇‍♂️

> @parasharrajat any update on my proposal ? Have you fixed this issue ?

**Comment by parasharrajat** [comment]
ETA: 4 hours.

**Comment by parasharrajat** [comment]
@aneequeahmad I am not sure where exactly you are referring to doing the change. Please update your proposal and use links to pinpoint the change. Also, do I have to add those lines or replace it?? :confused:

**Comment by aneequeahmad** [comment]
@parasharrajat updated my [proposal](https://github.com/Expensify/App/issues/8292#:~:text=SOLUTION%2002-,we%20can%20use%20ref%20to%20focus%20the%20input%20if%20we%20don%27t%20want%20to%20use%20state%20variable%20to%20set%20focus,-.%20We%20can%20simply). The change is in `componentDidMount` as updated in proposal. 
You've to add these lines.

**Comment by parasharrajat** [comment]
**Update:** I missed the slack discussion and there is no update on it. I will ping it back and see if we get some insight. https://github.com/Expensify/App/issues/8292#issuecomment-1142413769

**Comment by aneequeahmad** [comment]
@parasharrajat 
This happens sometimes and sometimes the UI is stuck on animation due to animation in process. 
> opening the report screen jumps the UI a bit and settled down

Have you tried using my `first solution` ? it fixes the issue for me.

**Comment by parasharrajat** [comment]
No, It is also not working. The reason is https://github.com/Expensify/App/issues/8292#issuecomment-1155104027. You may have missed that.

**Comment by aneequeahmad** [comment]
### PROPOSAL

**Problem**
As we know field is `auto-focused` before the animation is completed and it's because
- `InteractionManager.runAfterInteractions()` is not working properly on web as `official react-native` documentation says [here](https://reactnative.dev/docs/interactionmanager#:~:text=InteractionManager.runAfterInteractions()%20is%20not%20working%20properly%20on%20web.%20It%20triggers%20immediately%20without%20waiting%20until%20the%20interaction%20is%20finished.)

### Solution

State variable should be used for slow navigator transitions and `this.setState({shouldAutFocus: true})` should be added at `line# 341` 

https://github.com/Expensify/App/blob/0b5ecd3df3b46cb34c3adf93e1b712f47a0adb17/src/pages/home/report/ReportActionCompose.js#L337-L341

- StackOverflow(as referred in [stackoverflow](https://stackoverflow.com/questions/36127185/how-to-use-interactionmanager-runafterinteractions-make-navigator-transitions-fa) or by facebook)
- Facebook(Reference : https://facebook.github.io/react-native/docs/performance.html#slow-navigator-transitions

and `autofocus` on `Composer` should be this state variable as without it it is set before the component is mount and cause issues on web. 


<img width="527" alt="Screenshot 2022-06-30 at 7 47 18 AM" src="https://user-images.githubusercontent.com/36359331/176582010-6b83c39f-8f63-40cf-81c0-f8b6256a6bdd.png">

**Comment by aneequeahmad** [comment]
The problem is field is auto-focused without waiting for interaction to finish. This is mentioned in `react-native` official [doc](https://reactnative.dev/docs/interactionmanager#:~:text=Note%3A%20InteractionManager.runAfterInteractions()%20is%20not%20working%20properly%20on%20web.%20It%20triggers%20immediately%20without%20waiting%20until%20the%20interaction%20is%20finished.)
`Note: InteractionManager.runAfterInteractions() is not working properly on web. It triggers immediately without waiting until the interaction is finished.`

@parasharrajat  I think i'm missing something. Could you please help me understand ?

**Comment by tgolen** [comment]
Not overdue. Still getting proposals.

**Comment by tgolen** [comment]
Dropping to weekly for now. Daily seems overkill.

**Comment by aneequeahmad** [comment]
### PROPOSAL

**Problem**

- As we already know animation of drawer is not completed and input is focused.

- Also, `ReportActionCompose.js` is already mounted on page load i-e `componentDidMount` is invoked as `route home "/" is navigated to "/r/reportId"` and whenever a new chat(default room) is opened autofocus isn't invoked since component is `already mounted` and doesn't run `componentDidMount` but instead lifecycle method `componentDidUpdate` runs(Attaching a video below to see console.logs for what i said above)

**Solution:**

- We need to `focus input` at that specific point when 

> Chat is new.
> Drawer state is changed from open to close.
> Wait for animation to end before focusing Composer.

- Add this `code` in `componentDidUpdate` to focus composer 

```
    if(_.size(this.props.reportActions) === 1 && !!prevProps.isDrawerOpen && !this.props.isDrawerOpen) {
            InteractionManager.runAfterInteractions(() =>  setTimeout(() => this.textInput.focus(), 500))
     }
 
```
before this line of code https://github.com/Expensify/App/blob/7a37c014b231522fc7139acf6715e8d63dd04a7e/src/pages/home/report/ReportActionCompose.js#L181-L186

Also we can remove this check `_.size(this.props.reportActions) === 1 ` from Composer autofocus
https://github.com/Expensify/App/blob/7a37c014b231522fc7139acf6715e8d63dd04a7e/src/pages/home/report/ReportActionCompose.js#L591

- _**Note:**_

Tried using `300ms` was working fine for `content off screen and autofocus issue` but sometimes was causing glitch. Maybe my emulator is a bit slow but the time `500ms` works perfectly fine in every case and it isn't feeling slow IMO.

- **Video after the fix**

https://user-images.githubusercontent.com/36359331/179407843-143ff7a4-054b-43b5-aa08-1cdb8d03aa89.mp4

**Comment by adelekennedy** [comment]
not overdue, more proposals are coming in!

**Comment by parasharrajat** [comment]
The old discussion went cold. I will ping back on it with the latest state of this issue today.

**Comment by tgolen** [comment]
@parasharrajat It looks like you never provided an update on this issue. Mind adding one?

**Comment by ntdiary** [comment]
# Proposal
react-navigation doesn't provide a way to monitor animation completion for mobile web in legacy mode.
Its animation is realized by changing the inline style with javascript.
1. can use `MutationObserver` to observe its style attribute and execute a callback function when the animation ends.
2. this function can store the sidebar animation status with `onyx`.
3. the `ReportActionCompose` can use this status for focus.

Here is the demo:
https://github.com/ntdiary/App/commit/31bede020db38ea70229474e19db56440aed95c1

https://user-images.githubusercontent.com/8579651/184501981-952a9692-bb29-470f-9b7b-97f3cf4c81f9.mp4

**Comment by mollfpr** [comment]
### Proposal

The `ReportActionCompose` auto focus value is come from `shouldFocusInputOnScreenFocus` and `_.size(this.props.reportActions) === 1`.

For native and mWeb, `shouldFocusInputOnScreenFocus` will be `false`. So other than native and mWeb, report action will focus on screen open. This means that native and mWeb will have report action auto focus if the report is newly created.

So the goal is to remove autofocus only for mWeb. My proposal is to create a new function that determines should be report action focus based on window dimension and size of report actions.

The function is called `canReportActionFocus`.
`src/libs/canReportActionFocus/index.js`
```
export default (isSmallScreenWidth, isReportActionCreated) => {
    if (isSmallScreenWidth) {
        return false;
    }

    return isReportActionCreated;
};
```

`src/libs/canReportActionFocus/index.native.js`
```
export default (isSmallScreenWidth, isReportActionCreated = false) => isReportActionCreated;
```

Now updating the `autoFocus` prop on `Composer` to call the function `canReportActionFocus`.
https://github.com/Expensify/App/blob/c060200dd225d9b3497e508131733051bece2a25/src/pages/home/report/ReportActionCompose.js#L597
```diff
+  autoFocus={this.shouldFocusInputOnScreenFocus || canReportActionFocus(this.props.isSmallScreenWidth, _.size(this.props.reportActions) === 1)}
```


### Result

Web

https://user-images.githubusercontent.com/25520267/184693700-09213970-5343-4ca0-b17f-d463c58cfd83.mov

mWeb Safari

https://user-images.githubusercontent.com/25520267/184693870-7ea27baf-2e53-411b-b978-9ae35538ff77.mov

**Comment by aneequeahmad** [comment]
@parasharrajat have you checked my [proposal](https://github.com/Expensify/App/issues/8292#issuecomment-1186553287)  ? I gave it 1 month ago

**Comment by anboia** [comment]
@parasharrajat this problem also happens on Safari desktop, when isSmallScreenWidth is true. Should we also remove the autofocus for this case?

**Comment by anboia** [comment]
## Proposal

We can use the `shouldFocusInputOnScreenFocus` variable to set the autofocus behavior at the constructor function:

<img width="1023" alt="Screen Shot 2022-08-16 at 21 47 56" src="https://user-images.githubusercontent.com/4477665/185010482-9b27e63b-3786-4d27-a53a-8cdbe1d25c9b.png">

this way we prevent autofocus on desktop safari (small screen width), and only focus on mobile native if it is a chat room without messages. 

---
Suppose we consider adding autofocus on web mobile when `isSmallScreenWidth` is `false`. In that case, the bug will not show up, but the iPad (on safari) will remove the chat header, which might seem a little odd. We can do it by adding an `else` condition as follow: 
```javascript
} else {
    // mobile web
    this.shouldFocusInputOnScreenFocus = _.size(props.reportActions) === 1 && !props.isSmallScreenWidth;
}
```
![Simulator Screen Shot - iPad mini (6th generation) - 2022-08-16 at 21 51 54](https://user-images.githubusercontent.com/4477665/185012286-e24fe135-2e1b-4a86-a29f-73e4e07dd724.png)
*with autoFocus*

![Simulator Screen Shot - iPad mini (6th generation) - 2022-08-16 at 21 53 05](https://user-images.githubusercontent.com/4477665/185012337-55ab3ac9-8400-4f86-a681-9a19b8d0ce56.png)
*without autoFocus*

**Comment by ntdiary** [comment]
Hi @parasharrajat , Do the following examples behave as expected ? ( build from `main` branch )
If not, it may be better to explicitly call `focus` when appropriate. 

Can you help confirm these scenarios? 😄

1. reopen `#admins`, no auto focus

https://user-images.githubusercontent.com/8579651/185193514-f1fb6461-9f45-49e1-8367-ecc3b9b97886.mp4

2. open `#admins`, log out and then log in, open `#admin`, no auto focus

https://user-images.githubusercontent.com/8579651/185193854-54858c9a-ea77-43f0-97d0-f601405cacfb.mp4

3. open `#admins` with skeleton view, no auto focus

https://user-images.githubusercontent.com/8579651/185196658-c930ae04-a50f-41f1-a16c-a549cea60bcf.mp4

**Comment by adelekennedy** [comment]
not overdue - still reviewing proposals

**Comment by tgolen** [comment]
@parasharrajat how are you feeling about the proposals?

**Comment by mollfpr** [comment]
Thank you for your feedback @parasharrajat 

Here's my updated proposal with more explanation.

### Updated Proposal

Current for auto focusing composer we depends on `this.shouldFocusInputOnScreenFocus` or `_.size(this.props.reportActions) === 1`.

`this.shouldFocusInputOnScreenFocus` will be true if the user device is a non-touchscreen device, if the device is a touchscreen device then the focus will be decided is the report has just been created or not. But this causes the same issue if the user open the app on the Safari desktop and opens the app on a small screen layout.

https://user-images.githubusercontent.com/25520267/186685923-9beb716d-2afb-414e-a4d6-df4ff8bead4a.mov


`_.size(this.props.reportActions) === 1` this to focus the composer when the report just created on a mobile device (decided here https://github.com/Expensify/App/issues/4678#issuecomment-904413983). But this causes an issue when accessing the app on iOS safari.

---

#### Solution
Solution for issue on Safari desktop when resizing to the small screen. We need to prevent the composer focus on the small screen when deciding by can use a touch screen.
```diff
+ this.shouldFocusInputOnScreenFocus = !props.isSmallScreenWidth && canFocusInputOnScreenFocus();
```

Now we need to handle the focus when the device is:
1. Mobile app (This should be **focusing** the composer when the report is just created)
2. mWeb iOS/Android (This should be **not focusing** the composer at any time)
3. Opening the app on Safari tablet in landscape mode (This should be **focusing** the composer when the report is just created)

> I don't think that there is a need to create a function canReportActionFocus. These can be directly handled in the component with conditions.

> Also, we don't use Platform API directly in the code. Check out https://github.com/expensify/app#platform-specific-file-extensions.

For the above cases, we can't depend on the `isSmallScreenWidth` on the component cause mobile and mWeb will have the same value and for specific platform case, we should use the platform-specific file extension.

On the web function, we need to check the screen size. Because we want to handle cases 2 and 3 properly. The small screen check will handle case 2 and case 3 will be decided by the report just created.
`src/libs/canReportActionFocus/index.js`
```
export default (isReportActionCreated, isSmallScreenWidth) => {
    if (isSmallScreenWidth) {
        return false;
    }

    return isReportActionCreated;
};
```

For native devices, autofocus will be decided by the report just created.
`src/libs/canReportActionFocus/index.native.js`
```
export default isReportActionCreated => isReportActionCreated;
```

---

#### Result

Web

https://user-images.githubusercontent.com/25520267/184693700-09213970-5343-4ca0-b17f-d463c58cfd83.mov

mWeb Safari

https://user-images.githubusercontent.com/25520267/184693870-7ea27baf-2e53-411b-b978-9ae35538ff77.mov

https://user-images.githubusercontent.com/25520267/185027664-2c9ad67b-d616-45dd-b195-3431cfbb93dd.mov

iPad Safari

https://user-images.githubusercontent.com/25520267/186698173-ff4abc2c-781b-40bc-8268-710fe802ef5d.mp4

**Comment by ntdiary** [comment]
> It seems all the videos are iOS but the issue is present in iOS-safari.

@parasharrajat yes, I think if we just need to disable `auto focus` on the mobile web, there are already some good proposals here. I just wanted to suggest that the above problems can also be fixed if needed. 🙂 

In addition, I found a solution that can monitor the transition of the drawer (use `useDrawerProgress`, `useCode`),
so that we can keep `auto focus` on the mobile web if we want to. At the same time, the above problems will be fixed (with explicitly calling `focus`).
Are you interested in taking a brief look?
```js
import React, {useState} from 'react';
import PropTypes from 'prop-types';
import {useDrawerProgress, useDrawerStatus} from '@react-navigation/drawer';
import {call, useCode} from 'react-native-reanimated';
import _ from 'underscore';
import getComponentDisplayName from '../libs/getComponentDisplayName';
import CONST from '../CONST';

const withDrawerProgressPropTypes = {
    progress: PropTypes.oneOf(_.values(CONST.SIDEBAR_TRANSITION)),
};

// it can fix another issue.
let prevReportID = null;

export default function withDrawerProgress(WrappedComponent) {
    const WightDrawerProgress = (props) => {
        const [progress, setProgress] = useState(null);

        let prevValue = 0;
        const drawerProgress = useDrawerProgress();
        useCode(() => call([drawerProgress], ([value]) => {
            // 0 means closed completely, we could set it to 0.01 (spring threshold), so that it can be triggered earlier.
            if (value <= 0.01) {
                setProgress(CONST.SIDEBAR_TRANSITION.CLOSED);
            } else if (value >= 1) {
                setProgress(CONST.SIDEBAR_TRANSITION.OPENED);
            } else if (value > prevValue) {
                setProgress(CONST.SIDEBAR_TRANSITION.OPENING);
            } else if (value < prevValue) {
                setProgress(CONST.SIDEBAR_TRANSITION.CLOSING);
            }
            prevValue = value;
        }), [drawerProgress]);

        const drawerStatus = useDrawerStatus();
        if (drawerStatus === 'closed') {
            prevReportID = props.route.params.reportID;
        } else if (progress === CONST.SIDEBAR_TRANSITION.OPENED) {
            prevReportID = props.route.params.reportID;
        }
        return (
            <WrappedComponent
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...props}
                ref={props.forwardedRef}
                reportId={prevReportID}
                progress={progress}
            />
        );
    };

    WightDrawerProgress.displayName = `withDrawerProgress(${getComponentDisplayName(WrappedComponent)})`;
    WightDrawerProgress.propTypes = {
        forwardedRef: PropTypes.oneOfType([
            PropTypes.func,
            PropTypes.shape({current: PropTypes.instanceOf(React.Component)}),
        ]),
        route: PropTypes.shape({
            /** Route specific parameters used on this screen */
            params: PropTypes.shape({
                /** The ID of the report this screen should display */
                reportID: PropTypes.string,
            }).isRequired,
        }).isRequired,
    };
    WightDrawerProgress.defaultProps = {
        forwardedRef: undefined,
    };
    return React.forwardRef((props, ref) => (
        // eslint-disable-next-line react/jsx-props-no-spreading
        <WightDrawerProgress {...props} forwardedRef={ref} />
    ));
}

export {
    withDrawerProgressPropTypes,
};
```

commit: https://github.com/ntdiary/App/commits/issue-8292-demo

video:

https://user-images.githubusercontent.com/8579651/187250227-5d2b1bfe-d6fc-4f3c-8a9a-1aa1b7ac28fa.mp4

**Comment by ntdiary** [comment]
> But it is a little complex solution for this problem. I like to get more eyes on this before moving forward.

Yeah. I was trying to find a way to monitor the drawer transition in [issue 9253](https://github.com/Expensify/App/issues/9253#issuecomment-1227479860) last week, and it also happens to come in handy here if needed. 🙂

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by ntdiary** [comment]
Eh, without hooks, how about using a class component?
```js
import React, {Component} from 'react';
import {DrawerProgressContext} from '@react-navigation/drawer';
import Animated from 'react-native-reanimated';
import CONST from '../CONST';

const {call} = Animated;

class DrawerProgress extends Component {
    constructor(props) {
        super(props);
        this.prevValue = 0;
        this.progress = null;
        this.setProgress = this.setProgress.bind(this);
    }

    setProgress(progress) {
        if (this.progress === progress) {
            return;
        }
        this.progress = progress;

        // TODO:del
        console.log(progress);
    }

    render() {
        return (
            <Animated.Code exec={call([this.context], ([value]) => {
                // spring threshold, can set it to 0.01 or 0.09, so that event can be triggered earlier.
                if (value <= 0) {
                    this.setProgress(CONST.SIDEBAR_TRANSITION.CLOSED);
                } else if (value >= 1) {
                    this.setProgress(CONST.SIDEBAR_TRANSITION.OPENED);
                } else if (value > this.prevValue) {
                    this.setProgress(CONST.SIDEBAR_TRANSITION.OPENING);
                } else if (value < this.prevValue) {
                    this.setProgress(CONST.SIDEBAR_TRANSITION.CLOSING);
                }
                this.prevValue = value;
            })}
            />
        );
    }
}

DrawerProgress.contextType = DrawerProgressContext;

export default DrawerProgress;

```

log:

https://user-images.githubusercontent.com/8579651/188797985-3c0adb41-522d-444b-9fa2-e07cb93d4b2f.mp4

**Comment by tgolen** [comment]
Where do we stand with the proposal on this? I like the idea of getting those events put into the upstream code, but what does that mean for this issue in the meantime? Are we going to place it on hold, or move forward with `DrawerProgress`? I actually really like the look of `DrawerProgress` so I'd be cool with moving forward on that while we wait to see if we can get those events added upstream. If they ever get added upstream, then we can come back here and remove `DrawerProgress` eventually.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by parasharrajat** [comment]
the other issue is on now put on HOLD https://github.com/Expensify/App/issues/9253#issuecomment-1270575168.

**Comment by parasharrajat** [comment]
Nope. https://www.upwork.com/jobs/Content-off-screen-and-keyboard-opens-when-navigating-default-rooms-8292-Expensify_~016d3e6e7bdef8f679
